### DISCUSSION TITLE
Recommended way to add or recreate connectors after Client construction?

### DISCUSSION THREAD

**Discussion by ryanberckmans** [wevm/wagmi#1822] [Q&A]
gm

I'd like to add certain connectors dynamically and/or recreate a connector with the same id but with new config in a previously-constructed `Client`.

For example, the `Client` might be initialized on page load with an `InjectedConnector`, and later, if the user provides an email address, an additional connector is added for an oauth wallet. One reason to do this is to reduce the bundle size of the initial page load. Another reason is that some third-party libraries require their connector to be reinstantiated when its config changes.

What's the recommended to do this?

**Comment by tmm** [comment]
Check out [`client.setState`](https://github.com/wagmi-dev/wagmi/blob/main/packages/core/src/client.ts#L189). Note that this isn't an official API so is subject to change.

**Reply by ryanberckmans** [reply]
Ah, thanks! I looked at the client API but hadn't dug into setState. I knew I wasn't the first person to want to add a connector dynamically.

**Reply by ryanberckmans** [reply]
What's the recommended way to recreate a connector? This can be required when a third-party connector requires reinstantiation to change its config.

Example:
```typescript
function recreateConnector(wagmiClient: Client, newConnector: Connector): void {
  // invariant: wagmiClient.connectors already contains a connector with
  // the same id as newConnector.id, but they are different instances.
  // We want to destroy the existing connector with this id and then add newConnector.

  // TODO destroy wagmiClient.connector with id == newConnector.id
  //   what's the recommended code to do this?
  //   what if the connector to be destroyed is currently active or loading?

  wagmiClient.setState(s => {  // Add the new connector using setState
    return Object.assign({}, s, {
      connectors: [...s.connectors, c.connector],
    });
  });
}
```
